lib/zhr_devs/submissions/task.ex

```diff
@@ -8,41 +8,100 @@ defmodule ZhrDevs.Submissions.Task do
+
   import Algae
 
   @derive Jason.Encoder
 
+  defmodule Language do
+    @moduledoc "The supported language in which task is written"
+    @derive Jason.Encoder
+    defdata(Uptight.Text.t())
+
+    alias Uptight.Text, as: T
+
+    def new!(binary), do: %Language{language: T.new!(binary)}
+
+    def from_raw!(%Language{language: language}) do
+      T.un(language)
+    end
+  end
+
+  defmodule Library do
+    @moduledoc "The libraries used in task"
+    @derive Jason.Encoder
+    defdata(Uptight.Text.t())
+
+    alias Uptight.Text, as: T
+
+    def new!(binary), do: %Library{library: T.new!(binary)}
+
+    def from_raw!(%Library{library: library}) do
+      T.un(library)
+    end
+  end
+
+  defmodule Integration do
+    @moduledoc false
+
+    alias Uptight.Text, as: T
+
+    @derive Jason.Encoder
+    defdata(Uptight.Text.t())
+
+    def new!(binary), do: %Integration{integration: T.new!(binary)}
+
+    def from_raw!(%Integration{integration: integration}) do
+      T.un(integration)
+    end
+  end
+
+  defimpl Jason.Encoder, for: [Language, Library, Integration] do
+    def encode(%struct{} = value, opts) do
+      value |> struct.from_raw!() |> Jason.Encode.string(opts)
+    end
+  end
+
   defdata do
     task_name :: Uptight.Text.t()
-    programming_language :: Uptight.Text.t()
-    integrations :: list(Uptight.Text.t()) \\ []
-    library_stack :: list(Uptight.Text.t()) \\ []
+    programming_language :: Language.t()
+    integrations :: list(Integration.t()) \\ []
+    library_stack :: list(Library.t()) \\ []
   end
 
-  @required_keys [:task_name, :programming_language]
+  @list_keys [:integrations, :library_stack]
 
-  @spec from_uri(binary()) :: __MODULE__.t()
-  def from_uri(task_from_uri) when is_binary(task_from_uri) do
+  @spec new!(binary()) :: __MODULE__.t()
+  def new!(task_from_uri) do
     task_from_uri
     |> URI.decode()
     |> Jason.decode!(keys: :atoms!)
     |> Enum.map(fn
-      {key, value} when key in @required_keys and not is_binary(value) ->
-        raise "Invalid structure"
-
-      {key, nil} ->
+      {key, nil} when key in @list_keys ->
         {key, []}
 
-      {key, value} when is_binary(value) ->
-        {key, T.new!(value)}
+      {:task_name, value} when is_binary(value) and value != "" ->
+        {:task_name, T.new!(value)}
+
+      {:programming_language, value} when value in @supported_technologies ->
+        {:programming_language, Language.new!(value)}
 
-      {key, values} when is_list(values) ->
-        {key, Enum.map(values, &T.new!/1)}
+      {:integrations, values} when is_list(values) ->
+        {:integrations, Enum.map(values, &Integration.new!/1)}
+
+      {:library_stack, values} when is_list(values) ->
+        {:library_stack, Enum.map(values, &Library.new!/1)}
+
+      _ ->
+        raise "Invalid structure"
     end)
     |> then(fn fields -> struct!(__MODULE__, fields) end)
   end
 
-  def to_uri(%__MODULE__{} = task) do
+  def from_raw!(%__MODULE__{} = task) do
```

Comment:
(But I don't think that constructing this structure from raw elixir structure makes sense, so I don't think that this module should have / export from_raw! at all. There simply is no *one raw type* that is obvious raw counterpart of this structure. I guess, a map with __struct__ tags stripped? But we should not be in a situation where something has stripped __struct__ tags. They should be encoded with perfect fidelity into JSON and there should be a way to decode them back by abusing the fact that structs are maps).

Reply:
> They should be encoded with perfect fidelity into JSON and there should be a way to decode them back by abusing the fact that structs are maps

If there is one test that is **very** important, it's this one:

```
describe "tripping JSON representation" do
  task = generate_task()
  there = Jason.encode!(task)
  back_again = generic_simple_value_decoder(there)
  assert task == back_again
```

Reply:
You can write generic decoder in a way similar to my best effort encoder.

The only difference is that your returns Result.new and raises when it can't process some term (like a tuple).